addons/web/static/tests/utils.js

```diff
@@ -644,3 +644,82 @@ class Contains {
 export async function contains(selector, options) {
     await new Contains(selector, options).run();
 }
+
+const stepState = {
+    expectedSteps: null,
+    deferred: null,
+    timeout: null,
+    steps: [],
+
+    clear() {
+        clearTimeout(stepState.timeout);
+        this.timeout = null;
+        this.deferred = null;
+        this.steps = [];
+        this.expectedSteps = null;
+    },
+
+    check() {
+        this.timeout = setTimeout(() => {
+            QUnit.config.current.assert.verifySteps(this.expectedSteps);
+            this.deferred.reject(
+                new Error(`Timeout while waiting for steps: "${this.expectedSteps}"`)
+            );
+            this.clear();
+        }, 2000);
+        if (
+            this.expectedSteps.length !== this.steps.length ||
+            !this.expectedSteps.every((s, i) => s === this.steps[i])
+        ) {
+            return;
+        }
+        QUnit.config.current.assert.verifySteps(this.expectedSteps);
+        this.deferred.resolve();
+        this.clear();
+    },
+};
+
+if (window.QUnit) {
+    QUnit.testDone(() => {
+        if (stepState.expectedSteps) {
+            QUnit.config.current.assert.verifySteps(stepState.expectedSteps);
+        }
+        stepState.clear();
+    });
+}
+
+/**
+ * Assert that a step has been executed. It should then be verified by calling
+ * `assertSteps`.
+ *
+ * @param {string} step
+ */
+export function step(step) {
+    stepState.steps.push(step);
+    QUnit.config.current.assert.step(step);
+    if (stepState.expectedSteps) {
+        stepState.check();
+    }
+}
+
+/**
+ * Wait for the given steps to be executed or for the timeout to be reached.
+ *
+ * @param {string[]} steps
+ */
+export function assertSteps(steps) {
+    if (stepState.expectedSteps) {
+        throw new Error("Already waiting for steps.");
+    }
+    stepState.expectedSteps = steps;
+    stepState.deferred = makeDeferred();
+    stepState.timeout = setTimeout(() => {
+        QUnit.config.current.assert.verifySteps(stepState.expectedSteps);
+        stepState.deferred.reject(
+            new Error(`Timeout while waiting for steps: "${stepState.expectedSteps}"`)
+        );
+        stepState.clear();
```

Comment:
To avoid duplication of `verifySteps`/deferred manipulation, I think we should have the timeout call check() with a `crashOnFail` param.

This also allows to avoid inconsistency between `verifySteps` possibly being true here and the deferred being rejected anyway.